### ISSUE TITLE
Support `$interpolation`?

### ISSUE THREAD

**Issue by Goncalerta** [Goncalerta/proc-quote#3]
Currently, this crate follows the syntax of [dtolnay/quote](https://github.com/dtolnay/quote), where the symbol `#` is used for interpolations. I wonder if this symbol was chosen because `$` would clash with `macro_rules!`. 

The (unstable) [libproc_macro::quote](https://github.com/rust-lang/rust/blob/74256630117fedf479edbd1145cac47c88036aec/src/libproc_macro/quote.rs) uses `$` rather than `#`. Would allowing this symbol as an alias to `#` be beneficial? Using `$` would allow a smoother transition to `libproc` once `quote` is stable.

**Comment by dtolnay** [comment]
The choice of `#` is handy when writing a proc macro that emits macro_rules macros. If you go with `$`, you will want to come up with some syntax for escaping them, which is not needed for `#`.

**Comment by Goncalerta** [comment]
Interesting point. It didn't occur to me that someone could want to write a proc macro that emits macro_rules.

**Comment by eddyb** [comment]
I mentioned the implementation of `proc_macro::quote!` in the original thread (https://github.com/dtolnay/request-for-implementation/issues/8#issuecomment-458727004).

What it does is allow using `$$` to escape `$`, which is *probably fine*, although I'd prefer `\$` myself.

@dtolnay How do you deal with `#[...]` attribute syntax? Does it work out by simply not overlapping with `#(...)` or any other interpolation?

**Comment by dtolnay** [comment]
The quote crate does not have any syntax for escaping anything, and I don't recall anyone asking for escaping in the past 2 years. It allows `#[...]` without needing an escape.

~For `\$` presumably you would also want an escape sequence for `\` so that a macro can emit literally `\$` as two tokens.~ Er, is the problem that `\` is not allowed in macro input?

```console
error: unknown start of token: \
 --> src/main.rs:2:25
  |
2 |     println!(stringify!(\));
  |                         ^
```

**Comment by eddyb** [comment]
Yeah, `\` would have to be officially introduced, and you probably can't even allow it at all without giving it some semantics in `macro_rules` (because you can't add those later, and change the behavior of code using `\`).

This is a bit sad, as we could've introduced `\` a while back, and properly supported higher-order macros even in cases where repetitions are used, but we didn't.

It overall makes me have no preference between using `#` and having to deal with `$$`.
